File: javascript/lab.js
#!/usr/bin/env node

////////////////////////////////////////////////////////////////////////////////
//
// Supabase Lab
//
// lab.py - Main processing module
//
// Copyright (c) 2024, 2025 Andy Warmack
// This file is part of Supabase Lab, licensed under the MIT License.
// See the LICENSE file in the project root for more information.
////////////////////////////////////////////////////////////////////////////////

const core = require('./core');
const backend = require('./backend');
const session = require('./session');
const dev = require('./dev');

core.Main.version = 'v0.0.7js';

////////////////////////////////////////////////////////////////////////////////
// command-line interface
////////////////////////////////////////////////////////////////////////////////

let exit_command = false;

const parse = async (line) => {

  function parse_dot_references(match) {
    let dot_ref = match.slice(1);
    let ref_value = core.Session.config[dot_ref] ?? match;
    if (typeof ref_value === 'boolean') {
      ref_value = String(ref_value);
    }
    if (dot_ref == 'password') {
      ref_value = '*'.repeat(ref_value.length);
    }
    return ref_value;
  }

  // prompt for email
  function email_prompt(prompt) {
    return new Promise((resolve) => {
        core.Session.rl.question(prompt, (email) => {
            resolve(email);
        });
    });
  }

  // prompt for password
  function password_prompt(prompt) {
    return new Promise((resolve) => {
        process.stdout.write(prompt);
        const psw = process.stdout.write;
        process.stdout.write = () => {};
        core.Session.rl.question(prompt, (password) => {
            process.stdout.write = psw;
            process.stdout.write('\n');
            resolve(password);
        });
    });
  }

// parse input
  core.Session.current_input = ''; // clear input for next prompt
  line = line.replace(/\.\w+/g, parse_dot_references);
  let command = line;
  let args = '';
  const m = line.match(/^(\S*)\s(.*)$/);
  if (m) {
    command = m[1];
    args = m[2];
  }

  switch (command) {
    // login
    case 'login': {
      let email = '';
      let password = '';
      if ('email' in core.Session.config && 'password' in core.Session.config) {
        email = core.Session.config['email'];
        password = core.Session.config['password'];
      }
      if (email == '') {
        email = await email_prompt('Email: ');
      }
      if (password == '' ) {
        password = await password_prompt('Password: ');
      }
      await backend.sign_in(email, password);
      core.update_prompt();
      break;
    }

    // exit
    case 'exit': {
      exit_command = true;
      core.Session.rl.close();
      break;
    }

    // print
    case 'print': {
      if (args) {
        core.writeln(args);
      } else {
        core.writeln();
      }
      break;
    }

    // logout
    case 'logout': {
      backend.sign_out();
      break;
    }

    // undocumented, for development purposes
    case 'debug': {
      dev.debug(args);
      break;
    }
    case 'dev': {
      await dev.dev(args);
      break;
    }

    // unknown command
    default: {
      core.writeln('?\n   login\n   exit\n   print [text]');
    }
  }
};

////////////////////////////////////////////////////////////////////////////////
// readline processing
////////////////////////////////////////////////////////////////////////////////

// get user input
core.Session.rl.on('line', async (line) => {
  await parse(line);
  core.show_prompt(core.Session.rl);
});

// handle special keys during input
core.Session.rl.input.on('keypress', (char, key) => {
  if (key.name === 'backspace') {
    core.Session.current_input = core.Session.current_input.slice(0, -1);
  } else if (key.ctrl && key.name === 'u') {
    core.Session.current_input = '';
  } else if ((key.name == 'up') || (key.name === 'down')) {
    core.Session.current_input = core.Session.rl.line;
  } else if (char && char.length === 1 && /^[\x20-\x7E]$/.test(char)) {
    core.Session.current_input += char;
  } else if (key.name === 'return') {
    // EOL
  }
});

// handle session closure
core.Session.rl.on('close', () => {
  if (!exit_command) {
    core.writeln();
  }
  session.exit_completely();
});

////////////////////////////////////////////////////////////////////////////////
// start here
////////////////////////////////////////////////////////////////////////////////

const main = async () => {
  
  // get default configuration
  await session.get_config(['default.cfg']);

  // check for additional config file specification
  if (process.argv.length > 2) {
    await session.get_config(process.argv.slice(2));
  }
 
  // show Supabase Lab information
  core.show_info();
  
  // get Supabase connection
  await backend.connect();
};

main();

File: python/lab.py
#!/usr/bin/env python3

################################################################################
#
# Supabase Lab
#
# lab.py - Main processing module
#
# Copyright (c) 2024, 2025 Andy Warmack
# This file is part of Supabase Lab, licensed under the MIT License.
# See the LICENSE file in the project root for more information.
################################################################################

import asyncio
import getpass
import re
import sys

try:
    from prompt_toolkit import PromptSession, Application
    from prompt_toolkit.output import ColorDepth
    from prompt_toolkit.patch_stdout import patch_stdout
    from prompt_toolkit.history import InMemoryHistory
    PROMPT_TOOLKIT_IMPORTED = True
except ImportError:
    PROMPT_TOOLKIT_IMPORTED = False

import backend
import core
import dev
import session

core.Main.version = 'v0.0.7py'

################################################################################
# check for required imports
################################################################################

def prompt_toolkit_imported():
    return PROMPT_TOOLKIT_IMPORTED

################################################################################
# command-line interface
################################################################################

async def prompt():
    # prompt_toolkit setup
    core.Session.prompt_app = Application()
    history = InMemoryHistory()
    input_session = PromptSession(color_depth=ColorDepth.MONOCHROME, history=history)

    # in-place substitutions with config values
    def parse_dot_references(match):
        dot_ref = match.group(0)[1:]
        ref_value = core.Session.config.get(dot_ref, match.group(0))
        if isinstance(ref_value, bool):
            ref_value = str(ref_value)
        if dot_ref == 'password': # don't show passwords in clear text when using the print command
            ref_value = '*' * len(ref_value)
        return ref_value

    # main input loop
    while core.Main.running:
        # prompt for auth/session input
        input_prompt = core.Session.prompt if core.Session.authenticated else core.Main.auth_prompt
        line = await input_session.prompt_async(input_prompt)
        line = re.sub(r'\.\w+', parse_dot_references, line)
        await parse(line)

        # detect terminated session
        try:
            await core.Session.supabase.auth.get_user()
        except: # pylint: disable=bare-except
            core.Session.authenticated = False

async def parse(line):
    command = line
    args = None
    m = re.search(r'^(\S*)\s(.*)$', line)
    if m is not None:
        command = m.group(1)
        args = m.group(2)

    # nothing
    if command.strip() == '':
        return

    # login
    elif command == 'login':
        email = ''
        password = ''
        if 'email' in core.Session.config and 'password' in core.Session.config:
            email = core.Session.config['email']
            password = core.Session.config['password']
        if email == '':
            print('Email: ', end='')
            email = input()
        if password == '':
            password = getpass.getpass('Password: ')
        await backend.sign_in(email, password)

    # exit
    elif command == 'exit':
        await session.exit_completely()

    # print
    elif command == 'print':
        if args is not None:
            print(args)
        else:
            print()

    # logout
    elif command == 'logout':
        await backend.sign_out()

    # undocumented, for development purposes
    elif command == 'debug':
        dev.debug(args)
    elif command == 'dev':
        await dev.dev(args)

    # unknown command
    else:
        print(f"{command}?")
        print('   login')
        print('   exit')
        print('   print [text]')

################################################################################
# start here
################################################################################

async def main():

    # set up initial supabase connection
    await backend.connect()

    # get user input
    try:
        await prompt()
    except EOFError:
        await session.exit_completely()
    except KeyboardInterrupt:
        await session.exit_completely()
    except Exception as e: # pylint: disable=broad-exception-caught
        core.handle_error('prompt()', e)

# get default configuration
session.get_config(['default.cfg'])

# check for additional config file specification
if len(sys.argv) > 1:
    session.get_config(sys.argv[1:])

# show Supabase Lab information
core.show_info()

# check Python dependencies
if backend.supabase_imported() and backend.realtime_imported() and prompt_toolkit_imported():

    # start command-line interface
    with patch_stdout():
        asyncio.run(main())

# provide dependency guidance
else:
    if not backend.supabase_imported():
        core.info_print('The Supabase client for Python is required:  https://github.com/supabase/supabase-py')
    if not backend.realtime_imported():
        core.info_print('The Realtime library is required:  https://github.com/supabase/realtime-py')
    if not prompt_toolkit_imported():
        core.info_print('The Prompt Toolkit library is required:  https://python-prompt-toolkit.readthedocs.io/')
    if not backend.requests_imported():
        core.info_print('The Requests library may be required: https://requests.readthedocs.io/')
